### PR TITLE
x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher: add device handling

### PR DESCRIPTION
**Pull request by efd6**
## What does this PR do?

Adds device support to the Azure AD entity analytics input provider.

## Why is it important?

Milestone task needed for security solution.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

-

### CODE REVIEW

File: x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go [hunk 1]
```diff
@@ -165,6 +165,21 @@ func (p *azure) runFullSync(inputCtx v2.Context, store *kvstore.Store, client be
...
+		for _, d := range state.devices {
+			p.publishDevice(d, state, inputCtx.ID, client, tracker)
+		}
+
+		end := time.Now()
+		p.publishMarker(end, end, inputCtx.ID, false, client, tracker)
+
+		tracker.Wait()
+	}
+
```
Comment (bhapas) [inline x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go:182]: Should we consider a debug log if the devices length is > 0 ? Just thinking about troubleshooting this part without any log info?

File: x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go [hunk 2]
```diff
@@ -321,7 +370,25 @@ func (p *azure) doFetch(ctx context.Context, state *stateStore, fullSync bool) (
...
+	// Expand device group memberships.
+	updatedDevices.ForEach(func(devID uuid.UUID) {
+		d, ok := state.devices[devID]
+		if !ok {
+			p.logger.Errorf("Unable to find device %q in state", devID)
+			return
+		}
+		d.Modified = true
+		if d.Deleted {
+			return
```
Comment (bhapas) [inline x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go:382]: A debug log to say control exited from `d.Deleted` block ?

File: x-pack/filebeat/input/entityanalytics/provider/azuread/azure_test.go [hunk 3]
```diff
@@ -36,13 +36,18 @@ func TestAzure_DoFetch(t *testing.T) {
...
 	}
+	var wantModifiedDevices collections.UUIDSet
+	for _, v := range mockfetcher.DeviceResponse {
+		wantModifiedDevices.Add(v.ID)
+	}
 
-	require.Equal(t, wantModifiedUsers.Values(), got.Values())
+	require.Equal(t, wantModifiedUsers.Values(), gotUsers.Values())
+	require.Equal(t, wantModifiedDevices.Values(), gotDevices.Values())
 }
```
Comment (bhapas) [inline x-pack/filebeat/input/entityanalytics/provider/azuread/azure_test.go:53]: A lot of conditional blocks are introduced in `x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go` but unit testing has been very limited. Can we add more unit tests here

File: x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/device.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+package fetcher
+
+import (
+	"github.com/google/uuid"
+
+	"github.com/elastic/beats/v7/x-pack/filebeat/input/entityanalytics/internal/collections"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+)
+
+// TODO: Implement fetchers for the registered owners and users
```
Comment (bhapas) [inline x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/device.go:14]: A github issue link to the TODO ?

File: x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/device.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	// Discovered indicates that this device was newly discovered. This does not
+	// necessarily imply the device was recently added in Azure Active Directory,
+	// but it does indicate that it's the first time the device has been seen by
+	// the input.
+	Discovered bool `json:"-"`
+	// Modified indicates that an attribute or group membership has been
+	// modified on this device.
+	Modified bool `json:"-"`
+	// Deleted indicates the device has been deleted.
+	Deleted bool `json:"deleted"`
```
Comment (bhapas) [inline x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/device.go:53]: Are `Discovered` `Modified` `Deleted` Mutually exclusive?

File: x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/device.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+// Merge will merge the attributes, owners, users and group memberships of
+// another Device instance into this Device. The IDs of the devices must match.
+func (d *Device) Merge(other *Device) {
+	if d.ID != other.ID {
+		return
```
Comment (bhapas) [inline x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/device.go:60]: Log statement that merge did not happen due to non-matching ids?

File: x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/group.go [hunk 7]
```diff
@@ -11,9 +11,11 @@ type MemberType int
 
 const (
 	// MemberUser is a user.
-	MemberUser MemberType = iota
+	MemberUser MemberType = iota + 1
```
Comment (bhapas) [inline x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/group.go:14]: Why `+1` ?

**Review by bhapas — COMMENTED** [review]
Looks Good.
 
Just a thought though : There are instances where a lot of code is duplicated between `users` `groups` and `devices`. Not sure if we can think of a solution that helps to add more entities in future in a simple way?

File: x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/group.go (hunk 7 above)
Comment (efd6) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/group.go:14]: This prevents accidental zero values from being valid.

File: x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go [hunk 8]
```diff
@@ -229,37 +258,47 @@ func (p *azure) runIncrementalUpdate(inputCtx v2.Context, store *kvstore.Store,
 // and enriching users with group memberships. If fullSync is true, then any
 // existing deltaLink will be ignored, forcing a full synchronization from
 // Azure Active Directory. Returns a set of modified users by ID.
-func (p *azure) doFetch(ctx context.Context, state *stateStore, fullSync bool) (collections.UUIDSet, error) {
-	var updatedUsers collections.UUIDSet
-	var usersDeltaLink string
-	var groupsDeltaLink string
+func (p *azure) doFetch(ctx context.Context, state *stateStore, fullSync bool) (updatedUsers, updatedDevices collections.UUIDSet, err error) {
+	var usersDeltaLink, devicesDeltaLink, groupsDeltaLink string
```
[the remark this answers is not in the document]
Comment (efd6) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go:262]: There won't be additional entities.

File: x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/group.go (hunk 7 above)
Comment (bhapas) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/group.go:14]: Should the comment reflect the info?

File: x-pack/filebeat/input/entityanalytics/provider/azuread/azure_test.go (hunk 3 above)
Comment (efd6) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/azure_test.go:53]: AFAICS the proportion of testing has remained essentially unchanged with this addition.

File: x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/device.go (hunk 5 above)
Comment (efd6) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/device.go:53]: Yes. I was thinking of using an enum, but this is what was already being used, so I retained the same approach.

File: x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/device.go (hunk 6 above)
Comment (efd6) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/device.go:60]: @taylor-swanson?

File: x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go (hunk 2 above)
Comment (efd6) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go:382]: @taylor-swanson?

File: x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/group.go (hunk 7 above)
Comment (efd6) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/group.go:14]: I doesn't add anything for the reader it's just standard defensive programming.

Comment (bhapas) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/group.go:14]: Yeah...  Just to let other programmers know why this addition happened :-) unless there is a common aura among the programmers :-)

Comment (efd6) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/group.go:14]: They have blame and it will link to this discussion.

Comment (andrewkroh) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/fetcher/group.go:14]: > unless there is a common aura among the programmers :-)

It's a common pattern. My rough summary...

- Use `iota` when you want the zero value to the default.
- Use `iota + 1` when you don't want the zero value to be a valid enum value.
- And [sometimes](https://github.com/search?q=path%3A*.go+%2Finvalid%5Cw%2B+%5Cw%2B+%3D+iota%2F&type=code) you will see an explicit invalid zero value like `invalidFooType FooType = iota`.

File: x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go [hunk 9]
```diff
@@ -394,6 +461,47 @@ func (p *azure) publishUser(u *fetcher.User, state *stateStore, inputID string,
 	client.Publish(event)
 }
 
+// publishDevice will publish a user document using the given beat.Client.
```
Comment (andrewkroh) [inline x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go:464]: Should that be `s/user document/device document/`?

File: x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go [hunk 10]
```diff
@@ -165,6 +165,21 @@ func (p *azure) runFullSync(inputCtx v2.Context, store *kvstore.Store, client be
 		tracker.Wait()
 	}
 
+	if len(state.devices) != 0 {
+		tracker := kvstore.NewTxTracker(ctx)
+
```
Comment (taylor-swanson) [inline x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go:170]: I'm wondering, do we want a single set of writer markers to bound both users and devices? If we want them separate, do we want to attach any sort of field to indicate what the write markers are associated with? (device or user).

Thoughts, @andrewkroh?

File: x-pack/filebeat/input/entityanalytics/provider/azuread/statestore.go [hunk 11]
```diff
@@ -87,6 +95,18 @@ func newStateStore(store *kvstore.Store) (*stateStore, error) {
...
+	if err = s.tx.ForEach(devicesBucket, func(key, value []byte) error {
+		var d fetcher.Device
+		if err = json.Unmarshal(value, &d); err != nil {
+			return fmt.Errorf("unable to unmarshal device from state: %w", err)
+		}
+		s.devices[d.ID] = &d
+
+		return nil
+	}); err != nil && !errIsItemNotFound(err) {
+		return nil, fmt.Errorf("unable to get users from state: %w", err)
```
Comment (taylor-swanson) [inline x-pack/filebeat/input/entityanalytics/provider/azuread/statestore.go:107]: s/users/devices

File: x-pack/filebeat/docs/inputs/input-entity-analytics.asciidoc [hunk 12]
```diff
@@ -165,6 +172,46 @@ Example user document:
 }
 ----
 
+Device documents will show the current state of the device.
+
+Example device document:
```
[the remark this answers is not in the document]
Comment (efd6) [reply x-pack/filebeat/docs/inputs/input-entity-analytics.asciidoc:177]: That's crazy, I have avery clear memory of having changed this to the JSON from one of the examples in the mock package.

File: x-pack/filebeat/input/entityanalytics/provider/azuread/azure_test.go (hunk 3 above)
Comment (efd6) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/azure_test.go:53]: Found an area that wanted tests added, added input state to make the current tests cover it. The remainder should be covered by integration tests when the integration is a implemented.

**Review by andrewkroh — APPROVED** [review]

File: x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go (hunk 10 above)
Comment (andrewkroh) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go:174]: Good point, either is acceptable to me. If it is two separate marks then someone who is querying would need to be know if it was associated to the users or devices.

Comment (efd6) [reply x-pack/filebeat/input/entityanalytics/provider/azuread/azure.go:174]: ISTM a single set makes more sense. devices are users are really pretty much the same kind of entity (at the moment more so, with the TODO additions less so).

**Review by taylor-swanson — APPROVED** [review]

**Review by andrewkroh — APPROVED** [review]

**Pull request merged**